test/framework/ownerreference_helpers.go

```diff
@@ -0,0 +1,260 @@
+		if _, ok := allAssertFuncs[v.Object.Kind]; !ok {
+			allErrs = append(allErrs, errors.New(fmt.Sprintf("kind %s does not have an associated ownerRef assertion function", v.Object.Kind)))
+			continue
+		}
+		if err := allAssertFuncs[v.Object.Kind](v.Owners); err != nil {
+			allErrs = append(allErrs, errors.Wrapf(err, "Unexpected ownerReferences for %s/%s", v.Object.Kind, v.Object.Name))
+		}
+	}
+	Expect(kerrors.NewAggregate(allErrs)).To(BeNil())
+}
+
+// Kind and GVK for types in the core API package.
+var (
+	clusterClassKind       = "ClusterClass"
+	clusterKind            = "Cluster"
+	machineKind            = "Machine"
+	machineSetKind         = "MachineSet"
+	machineDeploymentKind  = "MachineDeployment"
+	machineHealthCheckKind = "MachineHealthCheck"
+
+	clusterClassGVK      = clusterv1.GroupVersion.WithKind(clusterClassKind)
+	clusterGVK           = clusterv1.GroupVersion.WithKind(clusterKind)
+	machineDeploymentGVK = clusterv1.GroupVersion.WithKind(machineDeploymentKind)
+	machineSetGVK        = clusterv1.GroupVersion.WithKind(machineSetKind)
+	machineGVK           = clusterv1.GroupVersion.WithKind(machineKind)
+)
+
+var CoreTypeOwnerReferenceAssertion = map[string]func([]metav1.OwnerReference) error{
+	clusterClassKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{})
+	},
+	clusterKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{})
+	},
+	machineDeploymentKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterGVK})
+	},
+	machineSetKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{machineDeploymentGVK})
+	},
+	machineKind: func(owners []metav1.OwnerReference) error {
+		return hasOneOfExactOwnersByGVK(owners, []schema.GroupVersionKind{machineSetGVK}, []schema.GroupVersionKind{kubeadmControlPlaneGVK})
+	},
+	machineHealthCheckKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterGVK})
+	},
+}
+
+// Kind and GVK for types in the exp package.
+var (
+	clusterResourceSetKind        = "ClusterResourceSet"
+	clusterResourceSetBindingKind = "ClusterResourceSetBinding"
+
+	clusterResourceSetGVK = addonsv1.GroupVersion.WithKind(clusterResourceSetKind)
+)
+var ExpOwnerReferenceAssertions = map[string]func([]metav1.OwnerReference) error{
+	clusterResourceSetKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{})
+	},
+	clusterResourceSetBindingKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterGVK, clusterResourceSetGVK})
+	},
+}
+
+var (
+	configMapKind = "ConfigMap"
+	secretKind    = "Secret"
+)
+
+var SecretOwnerReferenceAssertions = map[string]func([]metav1.OwnerReference) error{}
```

Comment:
Might be I'm wrong, but I was expecting Secrets to be owned by Cluster or ControlPlane (e.g CA, Kubeconfig) or by machines (Boostrap secrets)

Reply:
Also, we should provide comments about what each assertion is testing